### ISSUE TITLE
recording chrome hits an assert in monitor_fd_for_mapping()

### ISSUE THREAD

**Issue by pasko** [mozilla/rr#2624]
Steps to reproduce (assuming google-chrome-stable is installed (debian), works with locally-built chromium too):
```
rr record /opt/google/chrome/chrome \
    --user-data-dir=/tmp/rr-userdata \
    --no-sandbox \
    --allow-sandbox-debugging \
    http://cnn.com
```
Fails with:
```
[FATAL /path/to/src/rr/src/record_syscall.cc:5189:monitor_fd_for_mapping() errno: EACCES]
 (task 4011208 (rec:4011208) at time 51834)
 -> Assertion `dir' failed to hold. Can't open fd directory /proc/4011208/fd
```
(Does not fail with about:blank, I did not check other websites)
Reproduces 100% of time for me.

I could not make a short repro in a few days. What I kept trying is creating a few process children and grandchildren, killing one process, while doing mmap() in another, trying to stay close to what `strace -f` showed. In the end it looks like rr missed the fact that a process was killed (stayed zombie in this case, with all /proc/pid/* owned by root). The assert depends on the order of task reaping, which I do not know how to influence.

The regression appeared after this commit: 4e10088344fba26eac97c6d76836d1baaba96cdd. 

Crash details:
```
[FATAL /path/to/src/rr/src/record_syscall.cc:5189:monitor_fd_for_mapping() errno: EACCES]
 (task 4011208 (rec:4011208) at time 51834)
 -> Assertion `dir' failed to hold. Can't open fd directory /proc/4011208/fd
Tail of trace dump:
{
  real_time:7687048.542638 global_time:51814, event:`SYSCALLBUF_RESET' tid:4011192, ticks:459898
}
{
  real_time:7687048.542702 global_time:51815, event:`SYSCALL: futex' (state:EXITING_SYSCALL) tid:4011289, ticks:8795690
rax:0x0 rbx:0xca rcx:0xffffffffffffffff rdx:0x0 rsi:0x89 rdi:0x7f785487f6d8 rbp:0x7f785407ffa0 rsp:0x7f785407fd80 r8:0x0 r9:0xffffffff r10:0x7f785487f580 r11:0x246 r12:0x0 r13:0x7f785487f688 r14:0x7f785487f6d8 r15:0x7f785487f4f0 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xca fs_base:0x7f7854880700 gs_base:0x0
}
{
  real_time:7687048.542922 global_time:51816, event:`SYSCALLBUF_FLUSH' tid:4011289, ticks:8797199
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'futex', ret:0x0, size:0x14 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'sendto', ret:0x108, size:0x10 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
}
{
  real_time:7687048.542935 global_time:51817, event:`SYSCALL: futex' (state:ENTERING_SYSCALL) tid:4011289, ticks:8797199
rax:0xffffffffffffffda rbx:0xca rcx:0xffffffffffffffff rdx:0x0 rsi:0x89 rdi:0x7f785487f6d8 rbp:0x7f785407ffa0 rsp:0x7f785407fd80 r8:0x0 r9:0xffffffff r10:0x7f785487f580 r11:0x246 r12:0x0 r13:0x7f785487f688 r14:0x7f785487f6d8 r15:0x7f785487f4f0 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xca fs_base:0x7f7854880700 gs_base:0x0
}
{
  real_time:7687048.542943 global_time:51818, event:`SYSCALLBUF_RESET' tid:4011289, ticks:8797199
}
{
  real_time:7687048.543121 global_time:51819, event:`SYSCALL: epoll_wait' (state:EXITING_SYSCALL) tid:4011222, ticks:575546
rax:0x1 rbx:0xe8 rcx:0xffffffffffffffff rdx:0x20 rsi:0x2528cee83d80 rdi:0x1a rbp:0x7fe7841abfa0 rsp:0x7fe7841abd80 r8:0x0 r9:0x7fe783faa620 r10:0xffffffff r11:0x246 r12:0x0 r13:0xffffffff r14:0x7fe782aade2e r15:0x7fe782aadcae rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xe8 fs_base:0x7fe7849ac700 gs_base:0x0
  { tid:4011222, addr:0x2528cee83d80, length:0x180 }
}
{
  real_time:7687048.543328 global_time:51820, event:`SYSCALLBUF_FLUSH' tid:4011222, ticks:576935
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'recvmsg', ret:0x70, size:0x2d8 }
  { syscall:'futex', ret:0x0, size:0x14 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'epoll_wait', ret:0x0, size:0x10 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'epoll_wait', ret:0x0, size:0x10 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'epoll_wait', ret:0x0, size:0x10 }
}
{
  real_time:7687048.543342 global_time:51821, event:`SYSCALL: epoll_wait' (state:ENTERING_SYSCALL) tid:4011222, ticks:576935
rax:0xffffffffffffffda rbx:0xe8 rcx:0xffffffffffffffff rdx:0x20 rsi:0x2528cee83d80 rdi:0x1a rbp:0x7fe7841abfa0 rsp:0x7fe7841abd80 r8:0x0 r9:0x7fe783faa620 r10:0xffffffff r11:0x246 r12:0x0 r13:0xffffffff r14:0x7fe782aaa57e r15:0x7fe782aaa3fe rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xe8 fs_base:0x7fe7849ac700 gs_base:0x0
}
{
  real_time:7687048.543350 global_time:51822, event:`SYSCALLBUF_RESET' tid:4011222, ticks:576935
}
{
  real_time:7687048.543409 global_time:51823, event:`SYSCALL: futex' (state:EXITING_SYSCALL) tid:4011223, ticks:8461388
rax:0xffffffffffffff92 rbx:0xca rcx:0xffffffffffffffff rdx:0x0 rsi:0x89 rdi:0x7fe783faa628 rbp:0x7fe7837aafa0 rsp:0x7fe7837aad80 r8:0x0 r9:0xffffffff r10:0x7fe783faa4d0 r11:0x246 r12:0x0 r13:0x7fe783faa5d8 r14:0x7fe783faa628 r15:0x7fe783faa440 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xca fs_base:0x7fe783fab700 gs_base:0x0
}
{
  real_time:7687048.543690 global_time:51824, event:`SYSCALLBUF_FLUSH' tid:4011223, ticks:8468713
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'futex', ret:0x0, size:0x14 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'sendto', ret:0x70, size:0x10 }
  { syscall:'sendto', ret:0x110, size:0x10 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'sendto', ret:0x1e0, size:0x10 }
  { syscall:'sendto', ret:0x110, size:0x10 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
}
{
  real_time:7687048.543702 global_time:51825, event:`SYSCALL: futex' (state:ENTERING_SYSCALL) tid:4011223, ticks:8468713
rax:0xffffffffffffffda rbx:0xca rcx:0xffffffffffffffff rdx:0x0 rsi:0x89 rdi:0x7fe783faa628 rbp:0x7fe7837aafa0 rsp:0x7fe7837aad80 r8:0x0 r9:0xffffffff r10:0x7fe783faa4d0 r11:0x246 r12:0x0 r13:0x7fe783faa5d8 r14:0x7fe783faa628 r15:0x7fe783faa440 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xca fs_base:0x7fe783fab700 gs_base:0x0
}
{
  real_time:7687048.543709 global_time:51826, event:`SYSCALLBUF_RESET' tid:4011223, ticks:8468713
}
{
  real_time:7687048.544054 global_time:51827, event:`SYSCALL: epoll_wait' (state:EXITING_SYSCALL) tid:4011139, ticks:8518935
rax:0x1 rbx:0xe8 rcx:0xffffffffffffffff rdx:0x20 rsi:0x138ae9537780 rdi:0x1a rbp:0x7f60a26dffa0 rsp:0x7f60a26dfd80 r8:0x0 r9:0x31c r10:0x9c4 r11:0x246 r12:0x0 r13:0x9c4 r14:0x7f60a19dfbf6 r15:0x7f60a19dfa76 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xe8 fs_base:0x7f60a2ee0700 gs_base:0x0
  { tid:4011139, addr:0x138ae9537780, length:0x180 }
}
{
  real_time:7687048.544336 global_time:51828, event:`SYSCALLBUF_FLUSH' tid:4011139, ticks:8524635
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'recvmsg', ret:0x1f8, size:0x460 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'futex', ret:0x1, size:0x14 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'write', ret:0x1, size:0x10 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'write', ret:0x1, size:0x10 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'epoll_wait', ret:0x2, size:0x28 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'recvmsg', ret:0x1e0, size:0x448 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'read', ret:0x1, size:0x11 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'epoll_wait', ret:0x0, size:0x10 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'epoll_wait', ret:0x0, size:0x10 }
}
{
  real_time:7687048.544350 global_time:51829, event:`SYSCALL: epoll_wait' (state:ENTERING_SYSCALL) tid:4011139, ticks:8524635
rax:0xffffffffffffffda rbx:0xe8 rcx:0xffffffffffffffff rdx:0x20 rsi:0x138ae9537780 rdi:0x1a rbp:0x7f60a26dffa0 rsp:0x7f60a26dfd80 r8:0x0 r9:0x55e74da993f8 r10:0x9bc r11:0x246 r12:0x0 r13:0x9bc r14:0x7f60a19dfd9e r15:0x7f60a19dfc1e rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xe8 fs_base:0x7f60a2ee0700 gs_base:0x0
}
{
  real_time:7687048.544357 global_time:51830, event:`SYSCALLBUF_RESET' tid:4011139, ticks:8524635
}
{
  real_time:7687048.545264 global_time:51831, event:`SYSCALLBUF_FLUSH' tid:4011183, ticks:156608825
  { syscall:'madvise', ret:0x0, size:0x10 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'sendto', ret:0x38, size:0x10 }
  { syscall:'sendto', ret:0x38, size:0x10 }
  { syscall:'sendto', ret:0x38, size:0x10 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'gettimeofday', ret:0x0, size:0x28 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'gettimeofday', ret:0x0, size:0x28 }
  { syscall:'write', ret:0x1, size:0x10 }
  { syscall:'fcntl', ret:0x8002, size:0x10 }
}
{
  real_time:7687048.545278 global_time:51832, event:`SYSCALL: mmap' (state:ENTERING_SYSCALL) tid:4011183, ticks:156608825
rax:0xffffffffffffffda rbx:0x9 rcx:0xffffffffffffffff rdx:0x3 rsi:0x80000 rdi:0x0 rbp:0x7f785b2c9fa0 rsp:0x7f785b2c9d80 r8:0x1c r9:0x0 r10:0x1 r11:0x246 r12:0xaaaaaaaaaaaaaaaa r13:0x80000 r14:0x0 r15:0x0 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x9 fs_base:0x7f785d50ed80 gs_base:0x0
}
{
  real_time:7687048.545286 global_time:51833, event:`SYSCALLBUF_RESET' tid:4011183, ticks:156608825
}
[FATAL /path/to/src/rr/src/log.cc:402:emergency_debug()] (session doesn't look interactive, aborting emergency debugging)
=== Start rr backtrace:
rr(_ZN2rr13dump_rr_stackEv+0x35)[0x5566c865eaba]
rr(_ZN2rr15notifying_abortEv+0x53)[0x5566c865ea80]
rr(_ZN2rr12FatalOstreamD1Ev+0x30)[0x5566c84ef8a4]
rr(+0x376b59)[0x5566c84efb59]
rr(_ZN2rr21EmergencyDebugOstreamD1Ev+0x62)[0x5566c84efd74]
rr(+0x3d0098)[0x5566c8549098]
rr(+0x3d0e04)[0x5566c8549e04]
rr(+0x3e78e1)[0x5566c85608e1]
rr(+0x3d3397)[0x5566c854c397]
rr(_ZN2rr19rec_process_syscallEPNS_10RecordTaskE+0xd9)[0x5566c854c4e4]
rr(_ZN2rr13RecordSession21syscall_state_changedEPNS_10RecordTaskEPNS0_9StepStateE+0xce5)[0x5566c852f125]
rr(_ZN2rr13RecordSession11record_stepEv+0x3f5)[0x5566c85345df]
rr(+0x3ae4ad)[0x5566c85274ad]
rr(_ZN2rr13RecordCommand3runERSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS7_EE+0x39c)[0x5566c8528008]
rr(main+0x20c)[0x5566c86791f3]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xeb)[0x7f88fcefee0b]
rr(_start+0x2a)[0x5566c842c60a]
=== End rr backtrace
```

**Comment by khuey** [comment]
Do you have f76fbc90f1b414cf007e97a525cf1f91d23fc5ed?

**Comment by pasko** [comment]
Yes: this is on a6b1ea69e938ac0461549df83655f8da7c363dd6, and:
```
> git merge-base HEAD f76fbc9
f76fbc90f1b414cf007e97a525cf1f91d23fc5ed
```

**Comment by rocallahan** [comment]
I can reproduce this!

**Comment by rocallahan** [comment]
So in my case I get
```
 (task 387453 (rec:387453) at time 58742)
 -> Assertion `dir' failed to hold. Can't open fd directory /proc/387453/fd
```
387453 is a thread of pid 387426. Task 387426 exited a long time ago in the trace:
```
{
  real_time:2054645.032495 global_time:47192, event:`SYSCALL: futex' (state:EXITING_SYSCALL) tid:387426, ticks:7949661
rax:0x0 rbx:0xca rcx:0xffffffffffffffff rdx:0x0 rsi:0x89 rdi:0x7ffe1b411018 rbp:0x7f057f0c2fa0 rsp:0x7f057f0c2d80 r8:0x0 r9:0xffffffff r10:0x7ffe1b410ed0 r11:0x246 r12:0
x0 r13:0x7ffe1b410fc8 r14:0x7ffe1b411018 r15:0x7ffe1b410e40 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xca fs_base:0x7f058deedd00 
gs_base:0x0
}
{
  real_time:2054645.032690 global_time:47193, event:`SIGNAL: SIGTERM(async)' tid:387426, ticks:7949661
rax:0x0 rbx:0xca rcx:0xffffffffffffffff rdx:0x0 rsi:0x89 rdi:0x7ffe1b411018 rbp:0x7f057f0c2fa0 rsp:0x7f057f0c2d80 r8:0x0 r9:0xffffffff r10:0x7ffe1b410ed0 r11:0x246 r12:0
x0 r13:0x7ffe1b410fc8 r14:0x7ffe1b411018 r15:0x7ffe1b410e40 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xca fs_base:0x7f058deedd00 
gs_base:0x0
}
{
  real_time:2054645.032820 global_time:47194, event:`SIGNAL_DELIVERY: SIGTERM(async)' tid:387426, ticks:7949661
rax:0x0 rbx:0xca rcx:0xffffffffffffffff rdx:0x0 rsi:0x89 rdi:0x7ffe1b411018 rbp:0x7f057f0c2fa0 rsp:0x7f057f0c2d80 r8:0x0 r9:0xffffffff r10:0x7ffe1b410ed0 r11:0x246 r12:0
x0 r13:0x7ffe1b410fc8 r14:0x7ffe1b411018 r15:0x7ffe1b410e40 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xca fs_base:0x7f058deedd00 
gs_base:0x0
  { tid:387426, addr:0x7f057f0c2d80, length:(nil) }
}
{
  real_time:2054645.032939 global_time:47195, event:`SCHED' tid:387426, ticks:7949661
rax:0x0 rbx:0xca rcx:0xffffffffffffffff rdx:0x0 rsi:0x89 rdi:0x7ffe1b411018 rbp:0x7f057f0c2fa0 rsp:0x7f057f0c2d80 r8:0x0 r9:0xffffffff r10:0x7ffe1b410ed0 r11:0x246 r12:0
x0 r13:0x7ffe1b410fc8 r14:0x7ffe1b411018 r15:0x7ffe1b410e40 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xca fs_base:0x7f058deedd00 
gs_base:0x0
}
{
  real_time:2054645.033018 global_time:47196, event:`EXIT' tid:387426, ticks:7949661
  { tid:387426, addr:0x7f058deedfd0, length:0x4 }
}
```
387453 is stuck in a futex syscall from even earlier:
```
{
  real_time:2054641.847648 global_time:30525, event:`SYSCALL: futex' (state:ENTERING_SYSCALL) tid:387453, ticks:3382
rax:0xffffffffffffffda rbx:0xca rcx:0xffffffffffffffff rdx:0x0 rsi:0x80 rdi:0x8d01d3be768 rbp:0x7f0578f99fa0 rsp:0x7f0578f99d80 r8:0x0 r9:0x0 r10:0x0 r11:0x246 r12:0x7f0
579799730 r13:0x8d01d3be630 r14:0x8d01d3be768 r15:0x0 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xca fs_base:0x7f057979a700 gs_bas
e:0x0
}
{
  real_time:2054641.847657 global_time:30526, event:`SYSCALLBUF_RESET' tid:387453, ticks:3382
}
```
and it has no other activity in the trace.

**Comment by rocallahan** [comment]
Both 387453 and 387462 are zombies, i.e. exited but not yet reaped So I guess one question is what happened to the PTRACE_EVENT_EXIT event for 387453.

**Comment by rocallahan** [comment]
Setting `RR_LOG=all` the Chrome run gets further but we crash on shutdown with the same assertion. The situation is similar; a task is a zombie, another task in the same threadgroup exited with SIGTERM, but rr hasn't received the PTRACE_EVENT_EXIT for the other task (or any other threads in that threadgroup, actually).

**Comment by rocallahan** [comment]
Doing that again, we die earlier before page rendering, so it is a bit random. Again, all tasks in the offending process are zombies, one got SIGTERM, we got PTRACE_EVENT_EXIT for other threads in the process, but for one, apparently we didn't :-(.

**Comment by rocallahan** [comment]
The log has no sign that we touched the offending task in any way, no PTRACE_CONT that could have raced with PTRACE_EVENT_EXIT or anything like that. The last touch is
```
[Scheduler] Trying task 393327 from yield queue
[Scheduler] Task event is SYSCALL: futex
[Scheduler]   393327 is blocked on SYSCALL: futex; checking status ...
[Task] waitpid(393327, NOHANG) returns 0
```

**Comment by rocallahan** [comment]
> The regression appeared after this commit: 4e10088.

Before that commit we would have marked all threads in the dying process as unstable and refused to proceed until we observed exits for all of them. Now, we can keep running other code and just process exit events for the threads of the dying process like other events.

Turns out that in each of these Chrome traces, not long before we hit the assertion, there is a SIGKILL sent to the process containing the rogue task. Looks like Chrome sends a SIGTERM to its process, rr's scheduling does not get around to scheduling this particular task of that process for a couple of seconds, then Chrome code notices the process hasn't exited and follows up with a SIGKILL. That is probably important.

**Comment by rocallahan** [comment]
Oh! Experiments show that if I SIGTERM something and hold it in the PTRACE_EVENT_EXIT stop, and send it SIGKILL, it leaves that ptrace stop and exits to become a zombie. (If I originally kill it with SIGKILL, a second SIGKILL has no effect.) Either I misremembered how this works, or it has changed.

Anyway, this explains the bug. The process gets SIGTERM, its tasks go into PTRACE_EVENT_EXIT stops, but for at least one of the tasks rr doesn't schedule it for a while and the process gets SIGKILL, which boots the task(s) out of the PTRACE_EVENT_EXIT stop, in which case rr never gets the exit notification, so the task is dead but rr doesn't know, and then we die in the assertion when we're trying to read the task's fds.

**Comment by rocallahan** [comment]
We could reduce the window for the bug to occur by changing the scheduler to prioritize tasks whose process we think is exiting ... kind of a return of the "unstable exit" idea but only for scheduling prioritization.

However, we probably just need to be more accepting that processes can disappear out from under us and tolerate that instead of dying in the assertion :-(.

**Comment by Keno** [comment]
> Either I misremembered how this works, or it has changed.

From what you're describing, it sounds like how it has always worked. We should try to be robust to the task just going away in more places. It's always a race.

**Comment by rocallahan** [comment]
It's hard to write a testcase for this :-(.

**Comment by rocallahan** [comment]
I did manage to write a testcase. It may be useful to extend it if we find other issues with tasks that exit unexpectedly.

**Comment by rocallahan** [comment]
The fix here isn't enough to get Chrome to work. We still crash when we get around to scheduling the already-dead task. That's easy to reproduce by extending the testcase here.

**Comment by rocallahan** [comment]
Fixed that in da7e2e80aeacf7b4d3efdf61ff310118f02ce51a. Chrome works now.

**Comment by pasko** [comment]
Omg that was fast! Thank you a lot!

On da7e2e80aeacf7b4d3efdf61ff310118f02ce51a I hit another assert:
```
[FATAL /path/to/src/rr/src/AutoRemoteSyscalls.cc:586:check_syscall_result()]
 (task 1321767 (rec:1321767) at time 87982)
 -> Assertion `false' failed to hold. Syscall mmap failed with errno EBADF
```

I'm not sure whether it is the new way the bug manifests itself or something unrelated happening later down the execution path. Have you seen this with the same repro steps? I can proceed with a new bug if you are interested in more details.

**Comment by rocallahan** [comment]
I did not see it with my "load cnn.com in Chrome" test; I was able to load CNN, scroll around, then quit the browser with no problems.

A stack trace of rr while it's stopped at that assertion would be helpful. Putting it in a new issue would be good. Thanks!
